// flatbuffers/index.js
export const SIZE_PREFIX_LENGTH = 4;
export var Encoding;
(function (Encoding) {
    Encoding[Encoding["UTF8_BYTES"] = 1] = "UTF8_BYTES";
})(Encoding || (Encoding = {}));
const UTF8_DECODER = new TextDecoder();
const UTF8_ENCODER = new TextEncoder();
export class ByteBuffer {
    dataView;
    rawBytes;
    bbPosition = 0;
    constructor(bytes) {
        this.rawBytes = bytes;
        this.dataView = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    }
    bytes() {
        return this.rawBytes;
    }
    position() {
        return this.bbPosition;
    }
    setPosition(position) {
        this.bbPosition = position;
    }
    readInt32(offset) {
        return this.dataView.getInt32(offset, true);
    }
    readUint8(offset) {
        return this.dataView.getUint8(offset);
    }
    readUint32(offset) {
        return this.dataView.getUint32(offset, true);
    }
    readUint64(offset) {
        return this.dataView.getBigUint64(offset, true);
    }
    readFloat32(offset) {
        return this.dataView.getFloat32(offset, true);
    }
    __offset(tableOffset, vtableOffset) {
        const vtable = tableOffset - this.readInt32(tableOffset);
        const vtableLength = this.readUint16(vtable);
        return vtableOffset < vtableLength ? this.readUint16(vtable + vtableOffset) : 0;
    }
    __indirect(offset) {
        return offset + this.readInt32(offset);
    }
    __vector(offset) {
        const vectorOffset = offset + this.readInt32(offset);
        return vectorOffset + SIZE_PREFIX_LENGTH;
    }
    __vector_len(offset) {
        const vectorOffset = offset + this.readInt32(offset);
        return this.readInt32(vectorOffset);
    }
    __string(offset, encoding) {
        const stringOffset = offset + this.readInt32(offset);
        const stringLength = this.readInt32(stringOffset);
        const stringBytes = this.rawBytes.subarray(stringOffset + SIZE_PREFIX_LENGTH, stringOffset + SIZE_PREFIX_LENGTH + stringLength);
        if (encoding === Encoding.UTF8_BYTES) {
            return stringBytes;
        }
        return UTF8_DECODER.decode(stringBytes);
    }
    __has_identifier(identifier) {
        if (identifier.length !== SIZE_PREFIX_LENGTH) {
            return false;
        }
        const identifierOffset = this.bbPosition + SIZE_PREFIX_LENGTH;
        if (identifierOffset + SIZE_PREFIX_LENGTH > this.rawBytes.length) {
            return false;
        }
        for (let index = 0; index < identifier.length; index += 1) {
            if (this.rawBytes[identifierOffset + index] !== identifier.charCodeAt(index)) {
                return false;
            }
        }
        return true;
    }
    readUint16(offset) {
        return this.dataView.getUint16(offset, true);
    }
}
export class Builder {
    bytesArray;
    dataView;
    spaceValue;
    minAlignValue = 1;
    vtable = null;
    objectStart = 0;
    vectorNumElems = 0;
    nested = false;
    forceDefaultsValue = false;
    constructor(initialSize = 1024) {
        const size = Math.max(initialSize, 1);
        this.bytesArray = new Uint8Array(size);
        this.dataView = new DataView(this.bytesArray.buffer);
        this.spaceValue = size;
    }
    clear() {
        this.spaceValue = this.bytesArray.length;
        this.minAlignValue = 1;
        this.vtable = null;
        this.objectStart = 0;
        this.vectorNumElems = 0;
        this.nested = false;
    }
    forceDefaults(forceDefaults) {
        this.forceDefaultsValue = forceDefaults;
    }
    bytes() {
        return this.asUint8Array();
    }
    asUint8Array() {
        return this.bytesArray.subarray(this.spaceValue);
    }
    dataBuffer() {
        return new ByteBuffer(this.asUint8Array());
    }
    offset() {
        return this.bytesArray.length - this.spaceValue;
    }
    createString(value) {
        const utf8 = UTF8_ENCODER.encode(value);
        this.addInt8(0);
        this.startVector(1, utf8.length, 1);
        this.placeBytes(utf8);
        return this.endVector();
    }
    createByteVector(value) {
        const bytes = value instanceof Uint8Array ? value : Uint8Array.from(value);
        this.startVector(1, bytes.length, 1);
        this.placeBytes(bytes);
        return this.endVector();
    }
    startObject(numFields) {
        this.assertNotNested("startObject");
        this.vtable = new Array(numFields).fill(0);
        this.objectStart = this.offset();
        this.nested = true;
    }
    addFieldOffset(voffset, value, defaultValue) {
        if (value === defaultValue && !this.forceDefaultsValue) {
            return;
        }
        this.addOffset(value);
        this.slot(voffset);
    }
    addFieldInt64(voffset, value, defaultValue) {
        if (value === defaultValue && !this.forceDefaultsValue) {
            return;
        }
        this.addInt64(value);
        this.slot(voffset);
    }
    addFieldInt32(voffset, value, defaultValue) {
        if (value === defaultValue && !this.forceDefaultsValue) {
            return;
        }
        this.addInt32(value);
        this.slot(voffset);
    }
    endObject() {
        if (!this.vtable || !this.nested) {
            throw new Error("flatbuffers.Builder.endObject called without an active object.");
        }
        this.addInt32(0);
        const tableOffset = this.offset();
        for (let fieldIndex = this.vtable.length - 1; fieldIndex >= 0; fieldIndex -= 1) {
            const fieldOffset = this.vtable[fieldIndex] !== 0 ? tableOffset - this.vtable[fieldIndex] : 0;
            this.addInt16(fieldOffset);
        }
        const objectSize = tableOffset - this.objectStart;
        const vtableSize = (this.vtable.length + 2) * 2;
        this.addInt16(objectSize);
        this.addInt16(vtableSize);
        const objectStartIndex = this.bytesArray.length - tableOffset;
        this.dataView.setInt32(objectStartIndex, this.offset() - tableOffset, true);
        this.vtable = null;
        this.nested = false;
        return tableOffset;
    }
    finish(offset, fileIdentifier, sizePrefixed = false) {
        const additionalBytes = SIZE_PREFIX_LENGTH
            + (fileIdentifier ? SIZE_PREFIX_LENGTH : 0)
            + (sizePrefixed ? SIZE_PREFIX_LENGTH : 0);
        this.prep(this.minAlignValue, additionalBytes);
        if (fileIdentifier) {
            if (fileIdentifier.length !== SIZE_PREFIX_LENGTH) {
                throw new Error("FlatBuffers file identifiers must be exactly 4 characters.");
            }
            for (let index = fileIdentifier.length - 1; index >= 0; index -= 1) {
                this.writeInt8(fileIdentifier.charCodeAt(index));
            }
        }
        this.addOffset(offset);
        if (sizePrefixed) {
            this.addInt32(this.offset());
        }
    }
    startVector(elemSize, numElems, alignment) {
        this.assertNotNested("startVector");
        this.vectorNumElems = numElems;
        this.prep(SIZE_PREFIX_LENGTH, elemSize * numElems);
        this.prep(alignment, elemSize * numElems);
        this.nested = true;
    }
    addOffset(offset) {
        this.prep(SIZE_PREFIX_LENGTH, 0);
        if (offset > this.offset()) {
            throw new Error("flatbuffers.Builder.addOffset received an offset that is ahead of the current builder position.");
        }
        this.writeInt32(this.offset() + SIZE_PREFIX_LENGTH - offset);
    }
    addFloat32(value) {
        this.prep(4, 0);
        this.writeFloat32(value);
    }
    addInt8(value) {
        this.prep(1, 0);
        this.writeInt8(value);
    }
    addInt32(value) {
        this.prep(4, 0);
        this.writeInt32(value);
    }
    addInt64(value) {
        this.prep(8, 0);
        this.writeInt64(value);
    }
    endVector() {
        if (!this.nested) {
            throw new Error("flatbuffers.Builder.endVector called without an active vector.");
        }
        this.nested = false;
        this.addInt32(this.vectorNumElems);
        return this.offset();
    }
    addInt16(value) {
        this.prep(2, 0);
        this.writeInt16(value);
    }
    slot(fieldIndex) {
        if (!this.vtable) {
            throw new Error("flatbuffers.Builder.slot called without an active object.");
        }
        this.vtable[fieldIndex] = this.offset();
    }
    prep(size, additionalBytes) {
        if (size > this.minAlignValue) {
            this.minAlignValue = size;
        }
        let alignSize = ((~(this.bytesArray.length - this.spaceValue + additionalBytes)) + 1) & (size - 1);
        while (this.spaceValue < alignSize + size + additionalBytes) {
            this.growByteBuffer();
            alignSize = ((~(this.bytesArray.length - this.spaceValue + additionalBytes)) + 1) & (size - 1);
        }
        this.pad(alignSize);
    }
    pad(byteCount) {
        for (let index = 0; index < byteCount; index += 1) {
            this.bytesArray[--this.spaceValue] = 0;
        }
    }
    placeBytes(bytes) {
        this.spaceValue -= bytes.length;
        this.bytesArray.set(bytes, this.spaceValue);
    }
    growByteBuffer() {
        if (this.bytesArray.length >= 0x40000000) {
            throw new Error("FlatBuffers builder cannot grow beyond 1GB in this local runtime.");
        }
        const oldBytes = this.bytesArray;
        const newBytes = new Uint8Array(oldBytes.length * 2);
        newBytes.set(oldBytes, newBytes.length - oldBytes.length);
        this.spaceValue += newBytes.length - oldBytes.length;
        this.bytesArray = newBytes;
        this.dataView = new DataView(newBytes.buffer);
    }
    assertNotNested(methodName) {
        if (this.nested) {
            throw new Error(`flatbuffers.Builder.${methodName} requires depth-first serialization order.`);
        }
    }
    writeInt8(value) {
        this.dataView.setInt8(--this.spaceValue, value);
    }
    writeInt16(value) {
        this.spaceValue -= 2;
        this.dataView.setInt16(this.spaceValue, value, true);
    }
    writeInt32(value) {
        this.spaceValue -= 4;
        this.dataView.setInt32(this.spaceValue, value, true);
    }
    writeInt64(value) {
        this.spaceValue -= 8;
        this.dataView.setBigInt64(this.spaceValue, value, true);
    }
    writeFloat32(value) {
        this.spaceValue -= 4;
        this.dataView.setFloat32(this.spaceValue, value, true);
    }
}
