// flatbuffers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = exports.ByteBuffer = exports.Encoding = exports.SIZE_PREFIX_LENGTH = void 0;
exports.SIZE_PREFIX_LENGTH = 4;
var Encoding;
(function (Encoding) {
    Encoding[Encoding["UTF8_BYTES"] = 1] = "UTF8_BYTES";
})(Encoding || (exports.Encoding = Encoding = {}));
const UTF8_DECODER = new TextDecoder();
class ByteBuffer {
    dataView;
    rawBytes;
    bbPosition = 0;
    constructor(bytes) {
        this.rawBytes = bytes;
        this.dataView = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    }
    bytes() {
        return this.rawBytes;
    }
    position() {
        return this.bbPosition;
    }
    setPosition(position) {
        this.bbPosition = position;
    }
    readInt32(offset) {
        return this.dataView.getInt32(offset, true);
    }
    readUint8(offset) {
        return this.dataView.getUint8(offset);
    }
    readUint32(offset) {
        return this.dataView.getUint32(offset, true);
    }
    readUint64(offset) {
        return this.dataView.getBigUint64(offset, true);
    }
    readFloat32(offset) {
        return this.dataView.getFloat32(offset, true);
    }
    __offset(tableOffset, vtableOffset) {
        const vtable = tableOffset - this.readInt32(tableOffset);
        const vtableLength = this.readUint16(vtable);
        return vtableOffset < vtableLength ? this.readUint16(vtable + vtableOffset) : 0;
    }
    __indirect(offset) {
        return offset + this.readInt32(offset);
    }
    __vector(offset) {
        const vectorOffset = offset + this.readInt32(offset);
        return vectorOffset + exports.SIZE_PREFIX_LENGTH;
    }
    __vector_len(offset) {
        const vectorOffset = offset + this.readInt32(offset);
        return this.readInt32(vectorOffset);
    }
    __string(offset, encoding) {
        const stringOffset = offset + this.readInt32(offset);
        const stringLength = this.readInt32(stringOffset);
        const stringBytes = this.rawBytes.subarray(stringOffset + exports.SIZE_PREFIX_LENGTH, stringOffset + exports.SIZE_PREFIX_LENGTH + stringLength);
        if (encoding === Encoding.UTF8_BYTES) {
            return stringBytes;
        }
        return UTF8_DECODER.decode(stringBytes);
    }
    __has_identifier(identifier) {
        if (identifier.length !== exports.SIZE_PREFIX_LENGTH) {
            return false;
        }
        const identifierOffset = this.bbPosition + exports.SIZE_PREFIX_LENGTH;
        if (identifierOffset + exports.SIZE_PREFIX_LENGTH > this.rawBytes.length) {
            return false;
        }
        for (let index = 0; index < identifier.length; index += 1) {
            if (this.rawBytes[identifierOffset + index] !== identifier.charCodeAt(index)) {
                return false;
            }
        }
        return true;
    }
    readUint16(offset) {
        return this.dataView.getUint16(offset, true);
    }
}
exports.ByteBuffer = ByteBuffer;
class Builder {
    notImplemented(method) {
        throw new Error(`flatbuffers.Builder.${method} is not implemented in this local runtime yet.`);
    }
    startObject(_numFields) {
        this.notImplemented("startObject");
    }
    addFieldOffset(_voffset, _value, _defaultValue) {
        this.notImplemented("addFieldOffset");
    }
    addFieldInt64(_voffset, _value, _defaultValue) {
        this.notImplemented("addFieldInt64");
    }
    addFieldInt32(_voffset, _value, _defaultValue) {
        this.notImplemented("addFieldInt32");
    }
    endObject() {
        return this.notImplemented("endObject");
    }
    finish(_offset, _fileIdentifier, _sizePrefixed) {
        this.notImplemented("finish");
    }
    startVector(_elemSize, _numElems, _alignment) {
        this.notImplemented("startVector");
    }
    addOffset(_offset) {
        this.notImplemented("addOffset");
    }
    addFloat32(_value) {
        this.notImplemented("addFloat32");
    }
    addInt8(_value) {
        this.notImplemented("addInt8");
    }
    endVector() {
        return this.notImplemented("endVector");
    }
}
exports.Builder = Builder;
