// flatbuffers/index.ts
export type Offset = number;

export const SIZE_PREFIX_LENGTH = 4;

export enum Encoding {
  UTF8_BYTES = 1,
}

const UTF8_DECODER = new TextDecoder();
const UTF8_ENCODER = new TextEncoder();

export class ByteBuffer {
  private readonly dataView: DataView;
  private readonly rawBytes: Uint8Array;
  private bbPosition = 0;

  constructor(bytes: Uint8Array) {
    this.rawBytes = bytes;
    this.dataView = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  }

  bytes(): Uint8Array {
    return this.rawBytes;
  }

  position(): number {
    return this.bbPosition;
  }

  setPosition(position: number): void {
    this.bbPosition = position;
  }

  readInt32(offset: number): number {
    return this.dataView.getInt32(offset, true);
  }

  readUint8(offset: number): number {
    return this.dataView.getUint8(offset);
  }

  readUint32(offset: number): number {
    return this.dataView.getUint32(offset, true);
  }

  readUint64(offset: number): bigint {
    return this.dataView.getBigUint64(offset, true);
  }

  readFloat32(offset: number): number {
    return this.dataView.getFloat32(offset, true);
  }

  __offset(tableOffset: number, vtableOffset: number): number {
    const vtable = tableOffset - this.readInt32(tableOffset);
    const vtableLength = this.readUint16(vtable);
    return vtableOffset < vtableLength ? this.readUint16(vtable + vtableOffset) : 0;
  }

  __indirect(offset: number): number {
    return offset + this.readInt32(offset);
  }

  __vector(offset: number): number {
    const vectorOffset = offset + this.readInt32(offset);
    return vectorOffset + SIZE_PREFIX_LENGTH;
  }

  __vector_len(offset: number): number {
    const vectorOffset = offset + this.readInt32(offset);
    return this.readInt32(vectorOffset);
  }

  __string(offset: number, encoding?: Encoding): string | Uint8Array {
    const stringOffset = offset + this.readInt32(offset);
    const stringLength = this.readInt32(stringOffset);
    const stringBytes = this.rawBytes.subarray(
      stringOffset + SIZE_PREFIX_LENGTH,
      stringOffset + SIZE_PREFIX_LENGTH + stringLength,
    );
    if (encoding === Encoding.UTF8_BYTES) {
      return stringBytes;
    }
    return UTF8_DECODER.decode(stringBytes);
  }

  __has_identifier(identifier: string): boolean {
    if (identifier.length !== SIZE_PREFIX_LENGTH) {
      return false;
    }
    const identifierOffset = this.bbPosition + SIZE_PREFIX_LENGTH;
    if (identifierOffset + SIZE_PREFIX_LENGTH > this.rawBytes.length) {
      return false;
    }
    for (let index = 0; index < identifier.length; index += 1) {
      if (this.rawBytes[identifierOffset + index] !== identifier.charCodeAt(index)) {
        return false;
      }
    }
    return true;
  }

  private readUint16(offset: number): number {
    return this.dataView.getUint16(offset, true);
  }
}

export class Builder {
  private bytesArray: Uint8Array;
  private dataView: DataView;
  private spaceValue: number;
  private minAlignValue = 1;
  private vtable: number[] | null = null;
  private objectStart = 0;
  private vectorNumElems = 0;
  private nested = false;
  private forceDefaultsValue = false;

  constructor(initialSize = 1024) {
    const size = Math.max(initialSize, 1);
    this.bytesArray = new Uint8Array(size);
    this.dataView = new DataView(this.bytesArray.buffer);
    this.spaceValue = size;
  }

  clear(): void {
    this.spaceValue = this.bytesArray.length;
    this.minAlignValue = 1;
    this.vtable = null;
    this.objectStart = 0;
    this.vectorNumElems = 0;
    this.nested = false;
  }

  forceDefaults(forceDefaults: boolean): void {
    this.forceDefaultsValue = forceDefaults;
  }

  bytes(): Uint8Array {
    return this.asUint8Array();
  }

  asUint8Array(): Uint8Array {
    return this.bytesArray.subarray(this.spaceValue);
  }

  dataBuffer(): ByteBuffer {
    return new ByteBuffer(this.asUint8Array());
  }

  offset(): number {
    return this.bytesArray.length - this.spaceValue;
  }

  createString(value: string): Offset {
    const utf8 = UTF8_ENCODER.encode(value);
    this.addInt8(0);
    this.startVector(1, utf8.length, 1);
    this.placeBytes(utf8);
    return this.endVector();
  }

  createByteVector(value: Uint8Array | number[]): Offset {
    const bytes = value instanceof Uint8Array ? value : Uint8Array.from(value);
    this.startVector(1, bytes.length, 1);
    this.placeBytes(bytes);
    return this.endVector();
  }

  startObject(numFields: number): void {
    this.assertNotNested("startObject");
    this.vtable = new Array(numFields).fill(0);
    this.objectStart = this.offset();
    this.nested = true;
  }

  addFieldOffset(voffset: number, value: Offset, defaultValue: Offset): void {
    if (value === defaultValue && !this.forceDefaultsValue) {
      return;
    }
    this.addOffset(value);
    this.slot(voffset);
  }

  addFieldInt64(voffset: number, value: bigint, defaultValue: bigint): void {
    if (value === defaultValue && !this.forceDefaultsValue) {
      return;
    }
    this.addInt64(value);
    this.slot(voffset);
  }

  addFieldInt32(voffset: number, value: number, defaultValue: number): void {
    if (value === defaultValue && !this.forceDefaultsValue) {
      return;
    }
    this.addInt32(value);
    this.slot(voffset);
  }

  endObject(): Offset {
    if (!this.vtable || !this.nested) {
      throw new Error("flatbuffers.Builder.endObject called without an active object.");
    }

    this.addInt32(0);
    const tableOffset = this.offset();

    for (let fieldIndex = this.vtable.length - 1; fieldIndex >= 0; fieldIndex -= 1) {
      const fieldOffset = this.vtable[fieldIndex] !== 0 ? tableOffset - this.vtable[fieldIndex] : 0;
      this.addInt16(fieldOffset);
    }

    const objectSize = tableOffset - this.objectStart;
    const vtableSize = (this.vtable.length + 2) * 2;
    this.addInt16(objectSize);
    this.addInt16(vtableSize);

    const objectStartIndex = this.bytesArray.length - tableOffset;
    this.dataView.setInt32(objectStartIndex, this.offset() - tableOffset, true);

    this.vtable = null;
    this.nested = false;
    return tableOffset;
  }

  finish(offset: Offset, fileIdentifier?: string, sizePrefixed = false): void {
    const additionalBytes =
      SIZE_PREFIX_LENGTH
      + (fileIdentifier ? SIZE_PREFIX_LENGTH : 0)
      + (sizePrefixed ? SIZE_PREFIX_LENGTH : 0);

    this.prep(this.minAlignValue, additionalBytes);

    if (fileIdentifier) {
      if (fileIdentifier.length !== SIZE_PREFIX_LENGTH) {
        throw new Error("FlatBuffers file identifiers must be exactly 4 characters.");
      }
      for (let index = fileIdentifier.length - 1; index >= 0; index -= 1) {
        this.writeInt8(fileIdentifier.charCodeAt(index));
      }
    }

    this.addOffset(offset);

    if (sizePrefixed) {
      this.addInt32(this.offset());
    }
  }

  startVector(elemSize: number, numElems: number, alignment: number): void {
    this.assertNotNested("startVector");
    this.vectorNumElems = numElems;
    this.prep(SIZE_PREFIX_LENGTH, elemSize * numElems);
    this.prep(alignment, elemSize * numElems);
    this.nested = true;
  }

  addOffset(offset: Offset): void {
    this.prep(SIZE_PREFIX_LENGTH, 0);
    if (offset > this.offset()) {
      throw new Error("flatbuffers.Builder.addOffset received an offset that is ahead of the current builder position.");
    }
    this.writeInt32(this.offset() + SIZE_PREFIX_LENGTH - offset);
  }

  addFloat32(value: number): void {
    this.prep(4, 0);
    this.writeFloat32(value);
  }

  addInt8(value: number): void {
    this.prep(1, 0);
    this.writeInt8(value);
  }

  addInt32(value: number): void {
    this.prep(4, 0);
    this.writeInt32(value);
  }

  addInt64(value: bigint): void {
    this.prep(8, 0);
    this.writeInt64(value);
  }

  endVector(): Offset {
    if (!this.nested) {
      throw new Error("flatbuffers.Builder.endVector called without an active vector.");
    }
    this.nested = false;
    this.addInt32(this.vectorNumElems);
    return this.offset();
  }

  private addInt16(value: number): void {
    this.prep(2, 0);
    this.writeInt16(value);
  }

  private slot(fieldIndex: number): void {
    if (!this.vtable) {
      throw new Error("flatbuffers.Builder.slot called without an active object.");
    }
    this.vtable[fieldIndex] = this.offset();
  }

  private prep(size: number, additionalBytes: number): void {
    if (size > this.minAlignValue) {
      this.minAlignValue = size;
    }

    let alignSize = ((~(this.bytesArray.length - this.spaceValue + additionalBytes)) + 1) & (size - 1);
    while (this.spaceValue < alignSize + size + additionalBytes) {
      this.growByteBuffer();
      alignSize = ((~(this.bytesArray.length - this.spaceValue + additionalBytes)) + 1) & (size - 1);
    }

    this.pad(alignSize);
  }

  private pad(byteCount: number): void {
    for (let index = 0; index < byteCount; index += 1) {
      this.bytesArray[--this.spaceValue] = 0;
    }
  }

  private placeBytes(bytes: Uint8Array): void {
    this.spaceValue -= bytes.length;
    this.bytesArray.set(bytes, this.spaceValue);
  }

  private growByteBuffer(): void {
    if (this.bytesArray.length >= 0x40000000) {
      throw new Error("FlatBuffers builder cannot grow beyond 1GB in this local runtime.");
    }

    const oldBytes = this.bytesArray;
    const newBytes = new Uint8Array(oldBytes.length * 2);
    newBytes.set(oldBytes, newBytes.length - oldBytes.length);
    this.spaceValue += newBytes.length - oldBytes.length;
    this.bytesArray = newBytes;
    this.dataView = new DataView(newBytes.buffer);
  }

  private assertNotNested(methodName: string): void {
    if (this.nested) {
      throw new Error(`flatbuffers.Builder.${methodName} requires depth-first serialization order.`);
    }
  }

  private writeInt8(value: number): void {
    this.dataView.setInt8(--this.spaceValue, value);
  }

  private writeInt16(value: number): void {
    this.spaceValue -= 2;
    this.dataView.setInt16(this.spaceValue, value, true);
  }

  private writeInt32(value: number): void {
    this.spaceValue -= 4;
    this.dataView.setInt32(this.spaceValue, value, true);
  }

  private writeInt64(value: bigint): void {
    this.spaceValue -= 8;
    this.dataView.setBigInt64(this.spaceValue, value, true);
  }

  private writeFloat32(value: number): void {
    this.spaceValue -= 4;
    this.dataView.setFloat32(this.spaceValue, value, true);
  }
}
